Kconfig: add CONFIG_BOARD_REVISION

mbolivar-nordic opened the pull request:
Propagate the board revision to Kconfig via the environment.

An alternative to: https://github.com/zephyrproject-rtos/zephyr/pull/44131

mbolivar-nordic committed:
Kconfig: add CONFIG_BOARD_REVISION

Propagate the board revision to Kconfig via the environment.

This is useful for application code to have access to for similar
reasons that CONFIG_BOARD is useful.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
To try it out, do something like this:

```
$ west build --cmake-only -p -b nrf9160dk_nrf9160@0.7.0 samples/hello_world/
$ grep BOARD_REVISION build/zephyr/.config
CONFIG_BOARD_REVISION="0.7.0"
```

stephanosio commented on arch/Kconfig:
```diff
@@ -846,6 +846,9 @@ config BOARD
 	  if not found we look for the linker file in
 	  soc/<arch>/<family>/<series>
 
+config BOARD_REVISION
+	def_string "$BOARD_REVISION"
```
would be nice to add `help` as well.

by the way, is `arch/Kconfig` really the right place to add this config? IMO, `boards/Kconfig` would be more suited for this purpose.

carlescufi replied:
Agreed, `arch/Kconfig` is not the right place for this. +1 for `boards/Kconfig`.

mbolivar-nordic replied:
This is being added right under the definition for CONFIG_BOARD. Are you saying we should move both of them to boards/? Moving just this one would not make sense. 

@nashif @stephanosio

stephanosio reviewed (commented):
If we are going to add this as a Kconfig symbol, this would indeed be the correct thing to do.

No strong opinion on whether this should be a Kconfig symbol or just a good old `#define BOARD_REVISION`; though, I am inclined to say Kconfig is better since you can potentially reuse this to set the defaults for other configs.

carlescufi reviewed (dismissed):
I like this better.

nordicjm reviewed (dismissed):
Much simpler and better!

nashif reviewed (changes requested):
what @stephanosio  says...

mbolivar-nordic committed:
Kconfig: move CONFIG_BOARD to boards/Kconfig

Moving this option to the subdirectory for boards might make it easier
to find, and will keep it next to some other board-related Kconfig
options set in the same file.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
@nashif I have addressed @stephanosio's comments, including adding a prep work patch moving CONFIG_BOARD to boards/Kconfig to keep everything together.

tejlmand commented on boards/Kconfig:
```diff
@@ -9,6 +9,14 @@ config BOARD
 	  soc/<arch>/<family>/<series>
 
+config BOARD_REVISION
+	def_string "$BOARD_REVISION"
```
@mbolivar-nordic a little late to the game.

I'm sure this is useful but did you notice the fact that the `BOARD` Kconfig setting is not passed through the environment, but set in the board's Kconfig.

The result of that is that the BOARD value in Kconfig != BOARD value passed to CMake, se for example:
https://github.com/zephyrproject-rtos/zephyr/blob/a57dee94d68f2aa0f1a6a07512f1ceab6248c982/boards/arm/nrf5340dk_nrf5340/Kconfig.defconfig#L8-L9

This means that user can fetch the revision for the board given with `-DBOARD=<board>@<rev>`, but the board available in Kconfig may not be the same board, for example:
`-DBOARD=nrf5340dk_nrf5340_cpuapp_ns@1.2.3` results in:
- `CONFIG_BOARD == nrf5340dk_nrf5340_cpuapp`
- `CONFIG_BOARD_REVISION == 1.2.3`

so not sure passing revision this way is wise if we don't cleanup and enforce that `CONFIG_BOARD` is identical to the board provided by the user with `-DBOARD`.

mbolivar-nordic replied:
> so not sure passing revision this way is wise if we don't cleanup and enforce that `CONFIG_BOARD` is identical to the board provided by the user with `-DBOARD`.

are you aware of any counterexamples? That would indeed seem like a bug in the board defconfig if CONFIG_BOARD ended up different than BOARD (without the revision), wouldn't you agree?

tejlmand replied:
> are you aware of any counterexamples? That would indeed seem like a bug in the board defconfig if CONFIG_BOARD ended up different than BOARD (without the revision), wouldn't you agree?

@mbolivar-nordic did you look at the code I linked ?

That is exactly the counter example you are requesting.

A user specifying `-DBOARD=nrf5340dk_nrf5340_cpuapp_ns@0.1.2` will have the BOARD = nrf5340dk_nrf5340_cpuapp_ns is CMake, but as BOARD = nrf5340dk_nrf5340_cpuapp in Kconfig, that is without `_ns` in Kconfig.

tejlmand reviewed (commented):
a late arrival comment.
FYI @stephanosio @nashif @carlescufi @mbolivar-nordic
